### PR TITLE
Update RedirectController.php

### PR DESCRIPTION
**Pull request by pchampio**
If a password is pass as a get param to `/r/{album}` / `/r/{album}/{photo}`. Unlocks all the albums that the password unlocks.

Implemented in this manner in order for:
`https://domain.com/r/recent?password=my_password` to work!
companion lychee_front PR: https://github.com/LycheeOrg/Lychee-front/pull/229

### CODE REVIEW

File: app/Http/Controllers/RedirectController.php [hunk 1]
```diff
@@ -4,21 +4,71 @@
...
+                return redirect('gallery#' . $albumid . '/' . $photoid);
+        }
+
+        /**
+         * Provided an password, add all the albums that the password unlocks
+         *
+         * @param string $password
+         */
+        private function unlockPasswordAlbum(string $password) {
+                $albums = Album::whereNotNull('password')->where('password', '!=', '')->get();
```
Comment (ildyria) [inline app/Http/Controllers/RedirectController.php]: No, this is way too dangerous.
You know the album ID, thus you should query the album, check if the password is Ok,

And if it is a the correct password, then we also do the check for the other albums.

It is a bit slower but this increases the difficulty of unlocking all the albums by bruteforce.

**Review by ildyria — CHANGES_REQUESTED** [review]
aside from the check which is a bit too wide, this looks fine. :+1:

File: app/Http/Controllers/RedirectController.php (hunk 1 above)
Comment (pchampio) [reply app/Http/Controllers/RedirectController.php]: Given a photo uploaded to a password protected album.
I want this photo to be available for people opening the smart album `recent` with the following link: `https://domain.com/r/recent?password=my_password`.
How should I implement this?

Comment (ildyria) [reply app/Http/Controllers/RedirectController.php]: So I assume that your `recent` smart album is public ?

Comment (pchampio) [reply app/Http/Controllers/RedirectController.php]: Yes

Comment (ildyria) [reply app/Http/Controllers/RedirectController.php]: Just to clarify, you want to add a password to your `recent` folder or you want your `recent` folder to display recent pictures from the "_unlocked_" folders?

Comment (pchampio) [reply app/Http/Controllers/RedirectController.php]: I want the `recent` smart album to display recent pictures from the "unlocked" folders (and public photos).
And, I only want provide one share link to the `recent` smart album.

I don't want to have a password to the `recent` folder. But if a `?password=` is provided to the `recent` smart album, then unlock the photos.

Comment (pchampio) [reply app/Http/Controllers/RedirectController.php]: > `recent` folder to display recent pictures from the "unlocked" folders?

`recent` smart album currently works this way, if you unlock a password protected album in a given session. Then when this session browse the `recent` smart album, he will see the pictures from the "unlocked" folders in the `recent` album.

Comment (ildyria) [reply app/Http/Controllers/RedirectController.php]: Hooo I forgot about that.

**Review by ildyria — APPROVED** [review]
LGTM

**Pull request merged**